torchtitan/experiments/vlm/datasets/mm_datasets.py

```diff
@@ -0,0 +1,422 @@
+                return token_id[0]
+
+            special_tokens = [
+                _get_special_token_id(token)
+                for token in (
+                    IMAGE_TOKEN,
+                    BEGIN_OF_IMAGE_TOKEN,
+                    END_OF_IMAGE_TOKEN,
+                )
+            ]
+
+            labels = torch.where(
+                torch.isin(labels, torch.tensor(special_tokens)), IGNORE_INDEX, labels
+            )
+
+            # No truncation here - let collator handle it
+
+            # Keep images as list
+            pixel_values = processed_images  # List of tensors
+
+            return {
+                "input_ids": input_ids,
+                "labels": labels,
+                "pixel_values": pixel_values,
+            }
+
+        except Exception as e:
+            logger.warning(f"Error processing sample: {e}")
+            return None
+
+    def __iter__(self):
+        while True:
+            for sample in self._get_data_iter():
+                try:
+                    processed = self._process_sample(sample)
+                    if processed is None:
+                        continue
+
+                    # Simply yield individual samples - DataLoader will handle batching
+                    self._sample_idx += 1
+                    yield processed
+
+                except Exception as e:
+                    logger.warning(f"Error in iteration: {e}")
+                    continue
+
+            if not self.infinite:
+                break
+            else:
+                self._sample_idx = 0
+
+    def _get_data_iter(self):
+        try:
+            # For streaming datasets, we don't need to check length
+            if not hasattr(self._data, "iterable_dataset"):
+                if isinstance(self._data, Dataset) and self._sample_idx == len(
+                    self._data
+                ):
+                    return iter([])
+
+            it = iter(self._data)
+
+            # Skip samples if needed
+            if self._sample_idx > 0:
+                for _ in range(self._sample_idx):
+                    next(it)
+
+            return it
+        except Exception as e:
+            logger.error(f"Error in _get_data_iter: {e}")
+            return iter([])
+
+    def load_state_dict(self, state_dict):
+        self._sample_idx = state_dict["sample_idx"]
+
+    def state_dict(self):
+        return {"sample_idx": self._sample_idx}
+
+
+def build_mm_dataloader(
+    dp_world_size: int,
+    dp_rank: int,
+    tokenizer: BaseTokenizer,
+    job_config: JobConfig,
+    infinite: bool = True,
+) -> ParallelAwareDataloader:
+    """Build a data loader for HuggingFace datasets."""
+    dataset_name = job_config.training.dataset
+    dataset_path = job_config.training.dataset_path
+    batch_size = job_config.training.local_batch_size
+    seq_len = job_config.training.seq_len
+    # TODO: config
+    max_images_per_batch = batch_size * 2
+    max_patch_per_image = 256
+    patch_size = 16
```

Comment:
Are these 2 parameters referring to N, and L in the diagram you show? If so, we could add inline comments to indicate that, especially we use "N, L D" notations in collator.